ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/ModelConformanceValidator.scala

```diff
@@ -24,116 +19,121 @@ import com.daml.ledger.participant.state.kvutils.{Conversions, Err}
-            )
-            .fold(
-              err =>
-                rejections.buildRejectionStep(
-                  transactionEntry,
-                  rejectionReasonForValidationError(err),
-                  commitContext.recordTime,
-                ),
-              _ => StepContinue[DamlTransactionEntrySummary](transactionEntry),
-            )
-        } catch {
-          case err: Err.MissingInputState =>
-            logger.warn(
-              "Model conformance validation failed due to a missing input state (most likely due to invalid state on the participant)."
-            )
+  private def consumeValidationResult(
+      validationResult: Result[Unit],
+      transactionEntry: DamlTransactionEntrySummary,
+      commitContext: CommitContext,
+      rejections: Rejections,
+  )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
+    try {
+      val stepResult = for {
+        contractKeyInputs <- transactionEntry.transaction.contractKeyInputs.left
+          .map(rejectionForKeyInputError(transactionEntry, commitContext.recordTime, rejections))
+        _ <- validationResult
+          .consume(
+            lookupContract(commitContext),
+            lookupPackage(commitContext),
+            lookupKey(contractKeyInputs),
+          )
+          .left
+          .map(error =>
             rejections.buildRejectionStep(
               transactionEntry,
-              RejectionReasonV0.Disputed(err.getMessage),
+              RejectionReasonV0.Disputed(error.msg),
               commitContext.recordTime,
             )
-        }
-      })
-  }
-
-  private def contractIsActive(
-      transactionEntry: DamlTransactionEntrySummary,
-      contractState: DamlContractState,
-  ): Boolean = {
-    val activeAt = Option(contractState.getActiveAt).map(parseTimestamp)
-    !contractState.hasArchivedAt && activeAt.exists(transactionEntry.ledgerEffectiveTime >= _)
+          )
+      } yield ()
+      stepResult.fold(identity, _ => StepContinue(transactionEntry))
+    } catch {
+      case err: Err.MissingInputState =>
+        logger.warn(
+          "Model conformance validation failed due to a missing input state (most likely due to invalid state on the participant)."
+        )
+        rejections.buildRejectionStep(
+          transactionEntry,
+          RejectionReasonV0.Inconsistent(err.getMessage),
+          commitContext.recordTime,
+        )
+    }
   }
 
-  // Helper to lookup contract instances. We verify the activeness of
-  // contract instances here. Since we look up every contract that was
+  // Helper to lookup contract instances. Since we look up every contract that was
   // an input to a transaction, we do not need to verify the inputs separately.
-  private def lookupContract(
-      transactionEntry: DamlTransactionEntrySummary,
-      commitContext: CommitContext,
+  // A contract may not be in the state only if it's an edge case around committer pruning.
```

Comment:
if the contract was archived and pruned on the committer.
```suggestion
  // Note that a contract may not be in the state only if it was archived and pruned on the committer.
  // Then, the participant able to produce such a transaction only by using a divulged contract that
  // appeared as active, because it didn't learn about the archival. On the other hand, using divulged
  // contracts for interpretation is deprecated.

Reply:
Done